### PyTorch

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/c6/PyTorch_logo_black.svg/1920px-PyTorch_logo_black.svg.png" width="700" style="background:white">

- https://pytorch.org/ 공식사이트
- https://pytorch.kr/ 한글 공식사이트
- https://tutorials.pytorch.kr/  튜토리얼만 한글화 사이트

#### PyTorch 설치
- CPU 설치
    - 간단하지만, 그래픽카드 가속 사용할 수 없음
    - 기본 파이썬 환경 설치

- GPU 설치
    - 가상환경 생성, 설치

##### CPU 설치
- https://pytorch.org/get-started/locally/ 나와있는 Note에서 해당 플랫폼 종류 선택, 명령어 복사 실행
```python
!pip install torch torchvision torchaudio
```

In [1]:
!pip install torch torchvision torchaudio

   ---------------------------------------- 0.0/197.9 MB ? eta -:--:--
   ---------------------------------------- 1.6/197.9 MB 7.6 MB/s eta 0:00:26
    --------------------------------------- 3.7/197.9 MB 9.5 MB/s eta 0:00:21
   - -------------------------------------- 5.5/197.9 MB 9.3 MB/s eta 0:00:21
   - -------------------------------------- 7.6/197.9 MB 9.4 MB/s eta 0:00:21
   - -------------------------------------- 9.7/197.9 MB 9.6 MB/s eta 0:00:20
   -- ------------------------------------- 11.0/197.9 MB 9.0 MB/s eta 0:00:21
   -- ------------------------------------- 13.4/197.9 MB 9.3 MB/s eta 0:00:20
   --- ------------------------------------ 15.5/197.9 MB 9.4 MB/s eta 0:00:20
   --- ------------------------------------ 17.8/197.9 MB 9.5 MB/s eta 0:00:19
   --- ------------------------------------ 19.4/197.9 MB 9.4 MB/s eta 0:00:19
   ---- ----------------------------------- 21.5/197.9 MB 9.4 MB/s eta 0:00:19
   ---- ----------------------------------- 23.1/197.9 MB 9.4 MB/

In [2]:
import torch

In [3]:
## 2.4.0 + cpu = CPU 용 설치
torch.__version__

'2.4.0+cpu'

##### PyTorch GPU 설치
- NVIDIA GPU에서만 가능, CUDA로 되어있기 때문
- GPU 설치가 가능한지 먼저 확인

```python
!nvidia-smi
```

In [4]:
!nvidia-smi

Tue Aug  6 12:31:52 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 536.23                 Driver Version: 536.23       CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce GTX 1650      WDDM  | 00000000:01:00.0  On |                  N/A |
| 50%   32C    P8               7W /  75W |   1007MiB /  4096MiB |      9%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

##### CUDA Version 12.2
- CUDA(Compute Unified Device Architecture) - NVIDIA에서 그래픽카드 가속을 위해 만든 기술
- https://ko.wikipedia.org/wiki/CUDA

##### CUDA 설치
- CUDA Toolkit 12.2
    - https://developer.nvidia.com/cuda-downloads
- cuDNN(Deep Neural Network)
    - CUDA Toolkit에 맞는 버전확인
    - https://developer.nvidia.com/rdp/cudnn-archive
    - https://developer.nvidia.com/cuda-12-2-2-download-archive?target_os=Windows&target_arch=x86_64&target_version=11 확인 후 다운로드(단, 로그인 필수)
    - 압축을 풀고 CUDA 경로(C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2)에 복사/붙여넣기 

#### GPU용 파이썬 가상환경 만들기
- GPU PyTorch는 설치 로그를 봐야하기 때문에 명령 프롬프트에서 진행

- 터미널 오픈
    - virtualenv 확인
    - pip install virtualenv
- 가상환경 활성화
    - cd gpu_env\Scripts\activate

- PyTorch GPU 설치
    - 파이토치 GetStared 사이트에서 확인 명령어
    - pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121


#### 파이토치 GPU 확인

In [1]:
import torch

In [2]:
## 2.4.0+cu121 --> CUDA 12.1 버전의 PyTorch
torch.__version__

'2.4.0+cu121'

In [3]:
## CUDA 쓸 수 있는지 확인
torch.cuda.is_available()

True

In [4]:
## 설치된 CUDA가 동작하는 그래픽카드 정보확인
torch.cuda.get_device_name(0)

'NVIDIA GeForce GTX 1650'

##### 이제부터 GPU 가속으로 파이토치(텐서플로 가능) 딥러닝을 할 수 있음